### PR TITLE
added SitemapLoader

### PR DESCRIPTION
**Pull request by aaishikdutta**
Relates to https://github.com/embedchain/embedchain/issues/46. Added a SitemapLoader class that iterates over a list of urls from an XML sitemap and internally calls the WebPageLoader to load the urls. It is paired with the WebPageChunker to produce embedding for an entire website in one go.

### CODE REVIEW

**Comment by gasolin** [comment]
@aaishikdutta  looks cool, try to run it but got error:

`AttributeError: 'SitemapLoader' object has no attribute 'create_chunks'`

might missing the `embedchain/chunkers/site_map.py`

**Comment by aaishikdutta** [comment]
@gasolin can you please paste your code so that I can reproduce. Thanks

**Comment by gasolin** [comment]
@aaishikdutta sorry false alert, I copied chunkers to SitemapLoader instead of WebPageChunker 😅

**Comment by aaishikdutta** [comment]
all good @gasolin. Let me know if you find any other issues in this.

**Comment by cachho** [comment]
Can you write an exception for paths that point to images? Some sitemaps contain them.

**Comment by aaishikdutta** [comment]
Hey @cachho are you referring to this syntax. Sorry, haven't worked with these sitemaps previously:

`<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"
        xmlns:image="http://www.google.com/schemas/sitemap-image/1.1">
  <url>
    <loc>https://example.com/sample1.html</loc>
    <image:image>
      <image:loc>https://example.com/image.jpg</image:loc>
    </image:image>
    <image:image>
      <image:loc>https://example.com/photo.jpg</image:loc>
    </image:image>
  </url>
  <url>
    <loc>https://example.com/sample2.html</loc>
    <image:image>
      <image:loc>https://example.com/picture.jpg</image:loc>
    </image:image>
  </url>
</urlset>`

**Comment by cachho** [comment]
> Hey @cachho are you referring to this syntax. Sorry, haven't worked with these sitemaps previously:
> 
> 

I haven't seen that one, I guess you're just taking the `<loc>` so that images would be excluded anyways, right?

No I'm talking about stuff like this: https://en.wikipedia.org/wiki/File:Streathamladyjayne.jpg
Where File is in the URL, because they can't be handled by the embedding model anyways, as far as I know.

But I see how that would be too individual. I might have to filter the sitemap myself.

**Comment by cachho** [comment]
forget what I said, this is fine.

**Review by deshraj — COMMENTED** [review]
Looks good. Can you please resolve the merge conflicts and run `make format` to the code based on repository guidelines? Thanks!

**Comment by cachho** [comment]
resolved the merge conflicts for you

**Review by cachho — CHANGES_REQUESTED** [review]
Please add config and run `make format lint`

**Comment by aaishikdutta** [comment]
> Looks good. Can you please resolve the merge conflicts and run `make format` to the code based on repository guidelines? Thanks!

updated @deshraj 


> Please add config and run `make format lint`

updated @cachho

File: embedchain/loaders/site_map.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+                each_load_data = web_page_loader.load_data(link)
+                # WebPageLoader returns a list with single element
+                # which is extracted and appended to the output list
+                # containing data for all pages
+                output.append(each_load_data[0])
+
+            return output
+
+        else:
+            raise response.raise_for_status()
```
Comment (deshraj) [inline embedchain/loaders/site_map.py:33]: You can make the code a lot simpler by doing something like this:

```python
import requests
from bs4 import BeautifulSoup
from embedchain.loaders.web_page import WebPageLoader


class SitemapLoader:
    def load_data(self, sitemap_url):
        """
        This method takes a sitemap URL as input and retrieves
        all the URLs to use the WebPageLoader to load content
        of each page.
        """
        output = []
        web_page_loader = WebPageLoader()
        response = requests.get(sitemap_url)
        response.raise_for_status()

        soup = BeautifulSoup(response.text, "xml")
        links = [link.text for link in soup.find_all("loc")]
        for link in links:
            each_load_data = web_page_loader.load_data(link)
            output.append(each_load_data)
        return [data[0] for data in output]
```

Make sure to run `make format` after making the change.

Comment (aaishikdutta) [reply embedchain/loaders/site_map.py]: updated @deshraj

File: README.md [hunk 2]
```diff
@@ -294,6 +294,13 @@ To supply your own QnA pair, use the data_type as `qna_pair` and enter a tuple.
 ```python
 app.add_local('qna_pair', ("Question", "Answer"))
 ```
+### Sitemap
+
+To add a XML site map containing list of all urls, use the data_type as `site_map` and enter the sitemap url. Eg:
```
Comment (deshraj) [inline README.md:299]:
```suggestion
To add a XML site map containing list of all urls, use the data_type as `sitemap` and enter the sitemap url. Eg:
```

File: README.md [hunk 3]
```diff
@@ -294,6 +294,13 @@ To supply your own QnA pair, use the data_type as `qna_pair` and enter a tuple.
...
+
+```python
+app.add('site_map', 'a_valid_sitemap_url/sitemap.xml')
```
Comment (deshraj) [inline README.md:302]:
```suggestion
app.add('sitemap', 'a_valid_sitemap_url/sitemap.xml')
```

File: embedchain/data_formatter/data_formatter.py [hunk 4]
```diff
@@ -39,6 +40,7 @@ def _get_loader(self, data_type, config):
             "qna_pair": LocalQnaPairLoader(),
             "text": LocalTextLoader(),
             "docx": DocxFileLoader(),
+            "site_map": SitemapLoader(),
```
Comment (deshraj) [inline embedchain/data_formatter/data_formatter.py:43]:
```suggestion
            "sitemap": SitemapLoader(),
```

File: embedchain/data_formatter/data_formatter.py [hunk 5]
```diff
@@ -60,6 +62,7 @@ def _get_chunker(self, data_type, config):
             "qna_pair": QnaPairChunker(config),
             "text": TextChunker(config),
             "docx": DocxFileChunker(config),
+            "site_map": WebPageChunker(config),
```
Comment (deshraj) [inline embedchain/data_formatter/data_formatter.py:65]:
```suggestion
            "sitemap": WebPageChunker(config),
```

**Review by deshraj — APPROVED** [review]
Great, thanks for adding.

**Pull request merged**